src/nodes/hypertable_modify.c

```diff
@@ -93,6 +95,276 @@ get_chunk_dispatch_states(PlanState *substate)
+		}
+	}
+	table_close(chunk_rel, AccessShareLock);
+	/*
+	 * we get attrno for SEGMENT BY column from uncompressed chunk. This
+	 * can be different from that of compressed chunk due to different
+	 * physical layouts of compressed and uncompressed chunks. Thus we
+	 * need to retreive correct attrno from compressed chunk tuple descriptor.
+	 */
+	curr_chunk = linitial(context->chunks);
+	comp_chunk = ts_chunk_get_by_id(curr_chunk->fd.compressed_chunk_id, true);
+	comp_chunk_rel = relation_open(comp_chunk->table_id, AccessShareLock);
+	tupleDesc = RelationGetDescr(comp_chunk_rel);
+	for (attno = 1; attno <= tupleDesc->natts; attno++)
+	{
+		Form_pg_attribute attribute = TupleDescAttr(tupleDesc, attno - 1);
+		if (attribute->attisdropped)
+			continue;
+		char *attributeName = NameStr(attribute->attname);
+		if (strcmp(attributeName, column_name) == 0)
+		{
+			*attrno = attno;
+			break;
+		}
+	}
+	table_close(comp_chunk_rel, AccessShareLock);
+	return !is_null;
+}
+
+/*
+ * Helper method which returns true if given column_name
+ * is configured as SEGMENT BY column in a compressed hypertable
+ */
+static bool
+is_segmentby_col(List *ht_compression_info, char *column_name)
+{
+	ListCell *lc;
+	foreach (lc, ht_compression_info)
+	{
+		FormData_hypertable_compression *fd = lfirst(lc);
+		if (namestrcmp(&fd->attname, column_name) == 0)
+		{
+			if (fd->segmentby_column_index > 0)
+				return true;
+			break;
+		}
+	}
+	return false;
+}
+
+/*
+ * Helper method which will build scan keys for SEGMENT BY column
+ * with index attritbute number from compressed chunk.
+ */
+static void
+build_scankeys(HypertableModifyState *context, char *column_name, Const *const_value, int strategy)
+{
+	Chunk *curr_chunk = NULL;
+	Chunk *comp_chunk = NULL;
+	Relation comp_chunk_rel;
+	TupleDesc tupleDesc;
+	int attno = 0;
+	Form_pg_attribute attr;
+	Datum value;
+	ScanKeyData *scankey = NULL;
+	TypeCacheEntry *tce;
+
+	curr_chunk = linitial(context->chunks);
+	comp_chunk = ts_chunk_get_by_id(curr_chunk->fd.compressed_chunk_id, true);
+	comp_chunk_rel = relation_open(comp_chunk->table_id, AccessShareLock);
+	tupleDesc = RelationGetDescr(comp_chunk_rel);
+	scankey = palloc0(sizeof(ScanKeyData));
+	value = (const_value ? const_value->constvalue : 0);
+	/*
+	 * make compiler happy by silencing this error
+	 * ‘attr’ may be used uninitialized in this function
+	 */
+	attr = TupleDescAttr(tupleDesc, 0);
+	for (attno = 1; attno <= tupleDesc->natts; attno++)
+	{
+		attr = TupleDescAttr(tupleDesc, attno - 1);
+		if (namestrcmp(&(attr->attname), column_name) == 0)
+			break;
+	}
+	tce = lookup_type_cache(attr->atttypid, TYPECACHE_EQ_OPR_FINFO | TYPECACHE_CMP_PROC_FINFO);
+	ScanKeyEntryInitialize(scankey,
+						   0, /* flags */
+						   attno,
+						   strategy,
+						   InvalidOid, /* No strategy subtype. */
+						   attr->attcollation,
+						   (strategy == BTEqualStrategyNumber ? tce->eq_opr_finfo.fn_oid :
+																tce->cmp_proc_finfo.fn_oid),
+						   value);
```

Comment:
Minor nit: formatting looks off here.

Reply:
Regarding formatting iam not sure, there is a github action about code formatting which will fail if formatting was wrong. So i will correct it if github action complains it.